### PR TITLE
Encode reserved characters in database names

### LINKED ISSUE #760: support mongodb databases with spaces in their names
MongoDB itself doesn't allow this, but @jdegoes says it should be supported and abstracted by the back-end (slamdata/slamdata#249). We currently have an issue with this on the front-end, as the default name for directories is "Untitled Folder", so if you create a new directory in the top level of a mount that corresponds to a database, and attempting to create files within this database fails.

Uploading data to one of these invalidly-named database/directories results in a 500 error for a null pointer exception captured by Jetty or entirely empty response. Likewise, attempting to read or query stuff from an invalidly-named database also results in 500 errors with an NPE.

### LINKED ISSUE #818: Allow move files to destination that have "." symbol
I'm trying to move something (file, notebook, dir) from `"/path"` to `"/.trash/path"` and got that 
```json
{
    "error": "com.mongodb.MongoCommandException: Command failed with error -1: 'error with target namespace: invalid db name' on server localhost:27017. The full response is { \"ok\" : 0.0, \"errmsg\" : \"error with target namespace: invalid db name\" }"
}
```

moving to `"/trash/path"` works, moving to `"/.smth/path"` doesn't work

### PR DESCRIPTION
**Pull request by mossprescott**
Fixes #760 and fixes #818.

Commonly-used reserved characters are mapped to less-common, unreserved chars:
(space -> `+`, `.` -> `~`), and everything else is mapped to a name prefixed
with `$`.

### CODE REVIEW

File: core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala [hunk 1]
```diff
@@ -25,54 +26,128 @@ import slamdata.engine.fs._
...
+    def name: Parser[String] =
+      char.* ^^ { _.mkString }
+
+    def char: Parser[String] =
+      "$"  ^^ κ("$dollar") |
+      "+"  ^^ κ("$plus") |
+      "~"  ^^ κ("$tilde") |
+      " "  ^^ κ("+") |
+      "."  ^^ κ("~") |
+      "/"  ^^ κ("$slash") |
```
Comment (sellout) [inline core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala]: I was going to suggest mapping `.` to `/`, like how OS X swaps `:` and `/` in pathnames depending on the context … but it looks like `/` is actually reserved already, so meh.

File: core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala [hunk 2]
```diff
@@ -25,54 +26,128 @@ import slamdata.engine.fs._
...
+      "$slash"   ^^ κ("/")  |
+      "$bslash"  ^^ κ("\\") |
+      "$quote"   ^^ κ("\"") |
+      "$times"   ^^ κ("*" ) |
+      "$less"    ^^ κ("<" ) |
+      "$greater" ^^ κ(">")  |
+      "$colon"   ^^ κ(":")  |
+      "$bar"     ^^ κ("|")  |
+      "$qmark"   ^^ κ("?")  |
+      ".".r
```
Comment (sellout) [inline core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala]: Could we eliminate this mirrored structure by making a List of pairs, `[("$", "$dollar"), (".", "~"), …]`, and then fold over that to get the parser/unparser?

Comment (mossprescott) [inline core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala]: Yes.

File: core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala (hunk 1 above)
Comment (mossprescott) [inline core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala]: Right, and `:` is reserved, too. I considered `_` but decided to leave that one unencoded. Now's the time to suggest alternatives, because we're pretty much committed to this scheme once anyone use it.

File: core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala [hunk 3]
```diff
@@ -25,54 +26,128 @@ import slamdata.engine.fs._
...
+
+    def seg = segChar.* ^^ { _.mkString }
+
+    def segChar =
       "\\."  ^^ κ(".") |
       "\\d"  ^^ κ("$") |
       "\\\\" ^^ κ("\\") |
-      "."    ^^ κ("/") |
-      ".".r
+      "[^.]".r
```
Comment (sellout) [inline core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala]: Same.

File: core/src/test/scala/slamdata/engine/physical/mongodb/collection.scala [hunk 4]
```diff
@@ -95,8 +144,45 @@ class CollectionSpec extends Specification with DisjunctionMatchers {
...
+  } yield d
+  def genFile: Gen[FileNode] =  for {
+    n <- genName
+    f <- Gen.const(FileNode(n))
+  } yield f
+
+  def genName: Gen[String] = Gen.nonEmptyListOf(
+    Gen.oneOf(
+      Gen.oneOf(List.range('a', 'z')),
+      Gen.oneOf("$./\\_~ *+-".toList))).map(_.mkString)
```
Comment (sellout) [inline core/src/test/scala/slamdata/engine/physical/mongodb/collection.scala]: Why not just use an arbitrary string here? We should be encoding everything right? And we can size it to fit in the limit. That way we’d also test Unicode, too.

Comment (mossprescott) [inline core/src/test/scala/slamdata/engine/physical/mongodb/collection.scala]: I suppose, but the goal was more to hit all the escapes in every combination than to try to test every possible string. I would still want to include these explicitly, even if we also included Arbitrary[String].

The length issue is a bit tricky, because encoding fails if the *encoded* length is greater than 63/120. A sequence like `>>>>>>` hits that pretty fast (which is a reason to think about shortening some of those names).


This is a fun way to spend time, though:
```
> use 💩
switched to db 💩
> db.🐍.save({a: 1})
2015-07-08T08:57:16.320-0700 SyntaxError: Unexpected token ILLEGAL
> db.foo.save({a: 1})
WriteResult({ "nInserted" : 1 })
> show databases
_trash     0.078GB
admin      0.078GB
flarp      0.078GB
local      0.078GB
test       0.953GB
test_tmp   0.953GB
test_tmp2  0.078GB
tmp        0.078GB
~foo       (empty)
~trash     0.078GB
αβ         0.078GB
💩         0.078GB
```

**Comment by mossprescott** [comment]
Updated.

File: core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala [hunk 5]
```diff
@@ -25,54 +26,112 @@ import slamdata.engine.fs._
...
+  private object DatabaseNameParser extends PathParser {
+    def name: Parser[String] =
+      char.* ^^ { _.mkString }
+
+    def char: Parser[String] = substitute(DatabaseNameEscapes) | ".".r
+
+    def apply(input: String): PathError \/ String = parseAll(name, input) match {
+      case Success(name, _) =>
+        if (name.length > 63)
+          -\/(InvalidPathError("database name too long (> 63 chars): " + name))
```
Comment (sellout) [inline core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala]: We say “chars” here, but “bytes” in Collection. Are they actually distinct? If not, we should use the correct term … and if it’s bytes, we should mention the encoding 😝

**Comment by sellout** [comment]
The only thing that’s still open is the comment @mossprescott made about us ending up stuck with whatever mapping we choose, and so maybe we should shorten the encoding to avoid eating up the allowed length? Any thoughts on that, @jdegoes?

**Comment by jdegoes** [comment]
Because of the limitation on length (a mongo quirk), it does make sense to choose a compact encoding; and again most important chars for readability are '.', '/', and ' '.

File: core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala (hunk 5 above)
Comment (mossprescott) [inline core/src/main/scala/slamdata/engine/physical/mongodb/collection.scala]: You're getting really good at noticing things that cross my mind and then I decide to let them slide :-P

This is in fact straight from the [MongoDB documentation](http://docs.mongodb.org/manual/reference/limits/#Namespace-Length), which says "namespace" names (db name, `.`, collection name) have a max length of "120 bytes", and the source, which checks that database names have a length of no more than 64 (not 63, actually) chars, but now that I look more closely, those are C++ `char`, which seem to be bytes.

As for the encoding, I suppose that's a Java driver question, but I can't find any clues there. In BSON, all strings are apparently UTF-8, so that would be my guess.

So this means I should encode with UTF-8 before checking the length. Great.

**Comment by mossprescott** [comment]
` ` and `.` each get encoded as a single char (see above). `/` should now work (the old implementation could not handle it in Path), but gets encoded to `$slash`.

I think we all agree shorter is better, but it's also desirable to keep this somewhat readable/intuitive, since these names will be exposed in MongoDB. For more, see discussion on #760.

Out of the below, the only "obvious" shortenings I can think of are `$lt`, `$gt`, and maybe `$quot`. I'm open to suggestions.

```scala
    " "  -> "+",
    "."  -> "~",
    "$"  -> "$dollar",
    "+"  -> "$plus",
    "~"  -> "$tilde",
    "/"  -> "$slash",
    "\\" -> "$bslash",
    "\"" -> "$quote",
    "*"  -> "$times",
    "<"  -> "$less",
    ">"  -> "$greater",
    ":"  -> "$colon",
    "|"  -> "$bar",
    "?"  -> "$qmark"
```

**Comment by sellout** [comment]
I like those obvious shortenings – they also match up with XML entities, so there’s precedent 😉

**Comment by garyb** [comment]
Would `$$` work for `$dollar`?  `$add` for `+`, `$mul` or `$star` for `*` perhaps.

**Comment by sellout** [comment]
In the same vein as @garyb’s suggestion, `+` → `$+` and `~` → `$~`.

**Comment by sellout** [comment]
Why does MongoDB restrict so many characters?

**Comment by mossprescott** [comment]
Right, and `$div` for `/` is actually what scala offered. I thought that was a little less than obvious in this context, but `$slash` sucks, anyway.

One more I can think of is `$esc` for `\`.

That reduces almost all of the interesting characters to `$` plus three chars or less. I think that's an improvement.

Adopting all of those suggestions leaves us with this:
```scala
    " "  -> "+",
    "."  -> "~",
    "$"  -> "$$",
    "+"  -> "$add",
    "~"  -> "$tilde",
    "/"  -> "$div",
    "\\" -> "$esc",
    "\"" -> "$quot",
    "*"  -> "$mul",
    "<"  -> "$lt",
    ">"  -> "$gt",
    ":"  -> "$colon",
    "|"  -> "$bar",
    "?"  -> "$qmark"
```

**Comment by mossprescott** [comment]
Hmm. Not sure I like `$~`, but I can't argue with the logic. If I I could think of a shorter name than `$dollar`, I would avoid `$$`, too. `$usd`? `$dol`? `$peso`? Arg.

Actually, about half of these characters are reserved only on Windows, but it seemed simpler to apply the same rules regardless and not worry about where Mongo is running. All of them are path delimiters in some sense, with ` `, `.`, `/`, `\`, and `"` all being kind of a pain in one way or other to use in a Unix file name.

**Comment by mossprescott** [comment]
Updated (hours ago). Arg.
